File: Src/java/engine-jackson/src/test/java/org/opencds/cqf/cql/engine/execution/LetClauseOutsideQueryContextTest.java
package org.opencds.cqf.cql.engine.execution;

import java.util.List;

import org.opencds.cqf.cql.engine.elm.execution.EquivalentEvaluator;
import org.testng.Assert;
import org.testng.annotations.Test;

public class LetClauseOutsideQueryContextTest extends CqlExecutionTestBase {
    @Test
    public void testInterval() {
        Context context = new Context(library);

        Object result = context.resolveExpressionRef("First Position of list").getExpression().evaluate(context);
        Assert.assertTrue(EquivalentEvaluator.equivalent(((List<?>)result).get(0), 1));

        result = context.resolveExpressionRef("Third Position of list With Same Name of Let As First").getExpression().evaluate(context);
        Assert.assertTrue(EquivalentEvaluator.equivalent(((List<?>)result).get(0), 3));
    }
}

File: Src/java/engine/src/jvmTest/kotlin/org/opencds/cqf/cql/engine/execution/LetClauseOutsideQueryContextTest.kt
package org.opencds.cqf.cql.engine.execution

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.elm.executing.EquivalentEvaluator
import org.opencds.cqf.cql.engine.runtime.List
import org.opencds.cqf.cql.engine.runtime.toCqlInteger

internal class LetClauseOutsideQueryContextTest : CqlTestBase() {
    @Test
    fun evaluate() {
        val results =
            engine.evaluate { library("LetClauseOutsideQueryContextTest") }.onlyResultOrThrow
        var value = results["First Position of list"]!!.value
        Assertions.assertTrue(
            EquivalentEvaluator.equivalent((value as List).elementAt(0), 1.toCqlInteger()).value ==
                true
        )

        value = results["Third Position of list With Same Name of Let As First"]!!.value
        Assertions.assertTrue(
            EquivalentEvaluator.equivalent((value as List).elementAt(0), 3.toCqlInteger()).value ==
                true
        )
    }
}
